tpm2/test/cipher/aes_gcm_test.c

```diff
@@ -0,0 +1,307 @@
+  unsigned char* plaintext  = NULL;
+  unsigned char* ciphertext = NULL;
+  unsigned char* decrypt    = NULL;
+
+  int key_len = 16;
+  size_t plaintext_len = 16;
+  size_t ciphertext_len = 0;
+  size_t decrypt_len = 0;
+  
+  key = calloc(key_len, 1);
+  plaintext = calloc(plaintext_len, 1);
+  
+  CU_ASSERT(aes_gcm_encrypt(key, key_len, plaintext,
+            plaintext_len, &ciphertext, &ciphertext_len) == 0);
+  CU_ASSERT(ciphertext_len = plaintext_len + GCM_IV_LEN + GCM_TAG_LEN);
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext,
+            ciphertext_len, &decrypt, &decrypt_len) == 0);
+  CU_ASSERT(decrypt_len == plaintext_len);
+  CU_ASSERT(memcmp(plaintext, decrypt, plaintext_len) == 0);
+
+  free(decrypt);
+  free(key);
+  free(plaintext);
+  free(ciphertext);
+
+  return;
+}
+
+void test_gcm_key_modification(void)
+{
+  unsigned char* key        = NULL;
+  unsigned char* plaintext  = NULL;
+  unsigned char* ciphertext = NULL;
+  unsigned char* decrypt    = NULL;
+
+  int key_len = 16;
+  size_t plaintext_len = 16;
+  size_t ciphertext_len = 0;
+  size_t decrypt_len = 0;
+
+  key = calloc(key_len, 1);
+  plaintext = calloc(plaintext_len, 1);
+
+  // verify encryption/decryption of base case
+  aes_gcm_encrypt(key, key_len, plaintext, plaintext_len,
+                  &ciphertext, &ciphertext_len);
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 0);
+
+  free(decrypt);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // modify a single key bit
+  key[0] ^= 1;
+
+  // verify key modification breaks decryption
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+
+  free(plaintext);
+  free(ciphertext);
+  free(key);
+}
+	    
+void test_gcm_tag_modification(void)
+{
+  unsigned char* key        = NULL;
+  unsigned char* plaintext  = NULL;
+  unsigned char* ciphertext = NULL;
+  unsigned char* decrypt    = NULL;
+
+  int key_len = 16;
+  size_t plaintext_len = 16;
+  size_t ciphertext_len = 0;
+  size_t decrypt_len = 0;
+  
+  key = calloc(key_len, 1);
+  plaintext = calloc(plaintext_len, 1);
+
+  // check encryption/decryption of base case
+  aes_gcm_encrypt(key, key_len, plaintext, plaintext_len,
+                  &ciphertext, &ciphertext_len);
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 0);
+
+  free(decrypt);
+  decrypt = NULL;
+  decrypt_len = 0;
+
+  // alter the last byte of the tag and verify decryption failure
+  ciphertext[ciphertext_len-1] ^= 0x1;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
+                            &decrypt, &decrypt_len) == 1);
+  decrypt_len = 0;
+  decrypt = NULL;
+ 
+  // truncate the tag by 2 bytes and verify decryption failure
+  ciphertext_len -= 2;
+  CU_ASSERT(aes_gcm_decrypt(key, key_len, ciphertext, ciphertext_len,
```

Comment:
If somehow the length wasn't failing, wouldn't this fail anyway because of:

    ciphertext[ciphertext_len-1] ^= 0x1;

Reply:
I think that the modified byte (last byte of ciphertext array is the last byte of tag) is one of the two bytes that is truncated by shortening the length, so it shouldn't fail because of that. Nonetheless, I think it probably makes sense to test the two more independently. Therefore, I reversed the order of the tests and only decremented the length (to truncate the tag) in the aes_gcm_decrypt() call itself.